File: testing/sample_project.js
import PanoramaClient from "../javascript-sdk/main.js";
import express from "express";
import fs from "fs";

const app = express();
const PORT = 3001;

async function main() {
  const client = new PanoramaClient();

  try {
    await client.init({
      api_key: "sample_api_key_1_0",
      id: 1,
    });

    app.listen(PORT, () => {
      console.log(`Sample project listening at http://localhost:${PORT}`);
    });

    app.get("/", (req, res) => {
      res.send(`
        Sample Project is active!
        <br/><br/>
        Following endpoints can be used to test different features with Panorama:
        <ul>
          <li><code>/error</code></li>
        </ul>
        `);
    });

    app.get("/error1", (req, res) => {
      // simulate undefined function calling error
      try {
        client.addBreadcrumb({
          message: "User clicked on error1 endpoint",
          source: "user",
          type: "info",
        });
        const obj = {};
        client.addBreadcrumb({
          message: "Object value: " + JSON.stringify(obj),
          source: "log",
          type: "debug",
        });
        client.addBreadcrumb({
          message: "Object is empty",
          source: "log",
          type: "warning"
        });
        obj.nonExistentFunction();
      } catch (error) {
        client.captureError(error);
      }
      res.send("Successfully captured error with Panorama client!");
    });

    app.get("/error2", (req, res) => {
      // simulate changing constant variable
      try {
        const constantValue = "I am constant";
        client.addBreadcrumb({
          message: "User clicked on error2 endpoint",
          source: "user",
          type: "info"
        });
        constantValue = "something";
      } catch (error) {
        client.captureError(error);
      }
      res.send("Successfully captured error with Panorama client!");

    });

    app.get("/error3", (req, res) => {
      // simulate file not found error
      try {
        client.addBreadcrumb({
          message: "User clicked on error3 endpoint",
          source: "user",
          type: "info"
        });
        client.addBreadcrumb({
          message: "Attempting to read file data",
          source: "log",
          type: "debug"
        })
        fs.readFile("not_real_file.txt", "utf-8", (err, data) => {
          if (err) {
            client.captureError(err);
            res.send("Successfully captured error with Panorama client!");

          } else {
            res.send("File content: " + data);
          }

        })
      } catch (error) {
        client.captureError(error);
      }
      res.send("Successfully captured error with Panorama client!");
    });

    app.get("/benchmark1", (req, res) => {
      // simulate benchmark for a function
      function fibonacci(n) {
        if (n <= 1) {
          return n;
        }
        return fibonacci(n - 1) + fibonacci(n - 2);
      }
      client.beginBenchmark({
        name: "Fibonacci benchmark",
        expected_duration: 180,
      });
      const result = fibonacci(35);
      const benchmark = client.endBenchmark("Fibonacci benchmark");

      res.send("Fibonacci result: " + result + ". Completed in " + benchmark.duration + " ms");

    });

    app.get("/benchmark2", (req, res) => {
      // benchmark for random number generation with different amounts;
      function genRandomNumbers(x) {
        let ranNums = [];
        for (let i = 0; i < x; i++) {
          ranNums.push(Math.random());
        }
        return ranNums;
      }

      client.beginBenchmark({
        name: "Random number generation benchmark",
        expected_duration: 33,
      });
      const result = genRandomNumbers(1000000 + Math.floor(Math.random() * 500000));
      const benchmark = client.endBenchmark("Random number generation benchmark");
      

      res.send("Generated " + result.length + " random numbers in " + benchmark.duration + " ms");

    })

    app.get("/benchmark3", (req, res) => {
      // benchmark for sorting algorithm
      function selectionSort(arr) {
        const n = arr.length;
        for (let i =0; i < n; i++) {
          let minIndex = i;
          for (let j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIndex]) {
              minIndex = j;
            }
          }
          const temp = arr[i];
          arr[i] = arr[minIndex];
          arr[minIndex] = temp;


        }
        return arr;
      }

      const randomArr = [];
      for (let i = 0; i < 20000; i++) {
        randomArr.push(Math.floor(Math.random() * 500));
      }
      client.beginBenchmark({
        name: "Selection sort benchmark",
        expected_duration: 200
      })
      const result = selectionSort(randomArr);
      const benchmark = client.endBenchmark("Selection sort benchmark");

      res.send("Selection sorted array of length " + result.length + " in " + benchmark.duration + " ms");
    })
    
  } catch (error) {
    console.error("Error initializing sample project: ", error);
  }
}

main();

File: testing/sample_project.py
import python_sdk
from flask import Flask
import requests
import random
app = Flask(__name__)

print("Successfully imported PanoramaClient")
client = python_sdk.PanoramaClient()
client.init({
    "id": 1,
    "api_key": "sample_api_key_1_0",
    "api_url": "http://localhost:3000"
})

@app.route("/")
def home():
    return """
        Sample Project is active!
        <br/><br/>
        Following endpoints can be used to test different features with Panorama:
        <ul>
          <li><code>/error</code></li>
        </ul>
        `
"""

@app.route("/error1")
def error1():
    # testing undefined function call error
    client.add_breadcrumb("User clicked error1 endpoint", "user", "info")
    data = {}
    client.add_breadcrumb("Data object created", "log", "info")
    client.add_breadcrumb("Data object is empty", "log", "warning")
    try:
        data["undefined_function"].test()
    except Exception as e:
        client.capture_error(e)
    return "Successfully captured error #1"

@app.route("/error2")
def error2():
    # testing division by zero error
    client.add_breadcrumb("User clicked error2 endpoint", "user", "info")
    try:
        num1 = 5
        client.add_breadcrumb("num1 variable created with value 5", "log", "debug")
        num2 = 0
        client.add_breadcrumb("num2 variable created with value 0", "log", "debug")
        result = num1 / num2
    except Exception as e:
        client.capture_error(e)
    return "Successfully captured error #2"

@app.route("/error3")
def error3():
    # testing file not foudn error
    client.add_breadcrumb("User clicked error3 endpoint", "user", "info")
    try:
        file_name = "non_existent_file.txt"
        client.add_breadcrumb(f"Attempting to open file: {file_name}", "log", "debug")
        with open(file_name, "r") as f:
            content = f.read()
    except Exception as e:
        client.capture_error(e)
    return "Successfully captured error #3"

@app.route("/benchmark1")
def benchmark1():
    # testing fibonacci function benchmark
    client.begin_benchmark("Fibonacci benchmark", 220)
    def fibonacci(n):
        if n <= 1:
            return n
        else:
            return fibonacci(n-1) + fibonacci(n - 2)
    result = fibonacci(30)
    benchmark = client.end_benchmark("Fibonacci benchmark")
    return f"Benchmark completed in {benchmark['duration']} ms"
    
@app.route("/benchmark2")
def benchmark2():
    # benchmarking creating 200000 random numbers
    client.begin_benchmark("Create random numbers", 95)
    nums = []
    for i in range(200000):
        nums.append(random.randint(1,10000))
    
    benchmark = client.end_benchmark("Create random numbers")
    return f"Benchmark completed in {benchmark["duration"]} ms"
    
@app.route("/benchmark3")
def benchmark3():
    # benchmarking sorting list of 50000 random numbers
    client.begin_benchmark("Sort random numbers", 45)
    nums = []
    for i in range(50000):
        nums.append(random.randint(1, 10000))
    nums.sort()
    benchmark = client.end_benchmark("Sort random numbers")
    return f"Benchmark completed in {benchmark["duration"]} ms"
        
    
if __name__ == "__main__":
    app.run(port = 3001, debug = True)